lib/iris/cube.py

```diff
@@ -758,6 +767,231 @@ def _is_single_item(testee):
     return isinstance(testee, str) or not isinstance(testee, Iterable)
 
 
+class CubeAttrsDict(MutableMapping):
+    """
+    A dict-like object for "Cube.attributes", which provides unified user access to
+    the combined cube 'local' and 'global' attributes, mimicking the behaviour of a
+    simple dict.
+
+    This supports all the regular methods of a 'dict',
+    However, a few things such as the detailed print (repr) are different.
+
+    In addition, the 'locals' and 'globals' properties provide specific access to local
+    and global attributes, as regular :class:`~iris.common.mixin.LimitedAttributeDict`s.
+
+    Notes
+    -----
+    For type testing, "issubclass(CubeAttrsDict, Mapping)" is True, but
+    "issubclass(CubeAttrsDict, dict)" is False.
+
+    Properties 'locals' and 'globals' are the two sets of cube attributes.  These are
+    both :class:`~iris.common.mixin.LimitedAttributeDict`.  The CubeAttrsDict object
+    contains *no* additional state of its own, but simply acts as a view on these two.
+
+    When reading (__getitem__, pop. popitem, keys, values etc), it contains all the
+    keys + values of both 'locals' and 'globals'.  When a key occurs in *both* 'locals'
+    and 'globals', the result is the local value.
+
+    When writing (__setitem__, setdefault, update, etc) to a key already present, the
+    existing entry in either 'locals' or 'globals' is updated.  If both are present,
+    'locals' is updated.
+
+    When writing to a new key, this generally updates 'locals'.  However, certain
+    specific names would never normally be 'data' attributes, and these are created as
+    'globals' instead.  These are determined by Appendix A of the
+    `_CF Conventions: https://cfconventions.org/` .
+    At present, these are : 'conventions', 'featureType', 'history', 'title'.
+
+    Examples
+    --------
+    .. doctest::
+        :hide:
+        >>> cube = Cube([0])
+
+        >>> cube.attributes.update({'history':'fresh', 'x':3})
+        >>> print(cube.attributes)
+        {'history': 'fresh', 'x': 3}
+        >>> print(repr(cube.attributes))
+        CubeAttrsDict(globals={'history': 'fresh'}, locals={'x': 3})
+
+        >>> cube.attributes['history'] += '+added'
+        >>> print(repr(cube.attributes))
+        CubeAttrsDict(globals={'history': 'fresh+added'}, locals={'x': 3})
+
+        >>> cube.attributes.locals['history'] = 'per-var'
```

Comment:
Could you make these attribute values more similar to what you would encounter in real life? I think including real-world examples in the docs strings are more helpful for the user to understand what is going on

Reply:
For the **local** test attribute, I'm not aware of a "common" usage.  I changed it to "mycode", but I don't really know how helpful that is.  
For the **global** test attribute, I can only one of the defined names conventions / featureType / history / title, so perhaps 'title' might be better ?  Not sure though.  Can do, if you think that is clearer.

Reply:
I have no idea how to improve this so I'm just going to merge